### DISCUSSION TITLE
rich formatting in custom widget

### DISCUSSION THREAD

**Discussion by nschloe** [Textualize/textual#1002] [Q&A] [answered]
I'm trying to display Rich-formatted text in a table. With Textual 0.2+, this code

```python
from textual.app import App, ComposeResult, RenderResult
from textual.widgets import Static
from textual.widget import Widget
from rich.table import Table


class Info(Widget):
    def render(self):
        table = Table(show_header=False, expand=True, box=None, padding=0)
        table.add_column(justify="left", no_wrap=True)
        table.add_column(justify="right", no_wrap=True)
        table.add_row("[b]abc[/]", "[green]xyz[/]")
        return table


class FooApp(App):
    CSS = """
    .box {
        width: 100%;
        border: white;
        height: 1;
    }
    """

    def compose(self) -> ComposeResult:
        yield Static("One [b]two[/] three", classes="box")
        yield Info(classes="box")


foo_app = FooApp()
foo_app.run()
```

gives

![screenshot](https://user-images.githubusercontent.com/181628/197566986-a55e9c38-e46a-4f39-8ff8-c1aab3d90cc4.png)

Formatting is respected in `Static`, but not in my own widget.

Any hints?

**Comment by davep** [comment, marked as answer]
Hi! I've checked with @willmcgugan and, as of now, Textual won't render markup that is within a renderable, but *will* render markup when the `render` method returns a string with markup in it. This is by design but we're going to have a rethink of this. Meanwhile, to make your code work as you were intending, you can use Rich's `Text.from_markup` on your string data. For example, taking your `Info.Render` method above:

```python
    def render(self):
        table = Table(show_header=False, expand=True, box=None, padding=0)
        table.add_column(justify="left", no_wrap=True)
        table.add_column(justify="right", no_wrap=True)
        table.add_row(
            Text.from_markup("[b]abc[/]"),
            Text.from_markup("[green]xyz[/]")
        )
        return table
```

That will of course also need a `from rich.text import Text` in your code too.

**Reply by davep** [reply]
To follow up, @willmcgugan has raised #1023.
